src/org/etools/j1939_84/controllers/part2/Part02Step07Controller.java

```diff
@@ -48,6 +61,102 @@
+                                //  part 2 response and the part 1 response, as PGN 65259 data is defined
+                                //  to be static values.
+                                if (p.left.isPresent()) {
+                                    // PGN 65259 -> ComponentIdentificationPacket
+                                    ComponentIdentificationPacket packet = p.left.get();
+                                    if(module.getFunction() == 0){
+                                        dsFunctionZeroPacket[0] = packet;
+                                    }
+                                    if(!packet.getComponentIdentification().equals( module.getComponentIdentification()))
+                                        addFailure("6.2.7.2.b - " + moduleName +
+                                                           " reported component identification as: " + packet.getComponentIdentification().toString() +
+                                                           ", Part 01 Step 09 reported it as: " + module.getComponentIdentification().toString());
+                                    }
+                            }, () -> {
+                                //6.2.7.2 Fail criteria:
+                                // a. Fail if any device does not support PGN 65259 with the engine running that supported PGN 65259 with the engine off in part 1.
+                                addFailure("6.2.7.2.a There are no positive responses to a DS Component ID request from "
+                                                   + moduleName);
+                            });
+                    //6.2.7.5 Warn criteria2 for OBD ECUs other than function 0:
+                    // a. Warn if Component ID not supported for the global query
+                    // in 6.2.7.3 with engine running.
+                    if(module.getFunction() != 0 &&
+                            globalPackets.stream()
+                                    //not supported means didn't respond?
+                                    .noneMatch(packet -> packet.getSourceAddress() == moduleAddress)){
+                        addWarning("6.2.7.5.a - " + moduleName + " did not provide a positive respond to global query while engine running");
+                    }
+                    // Filter the modules responded to be only ones with function = 0 (engine function)
+                    // needed for 6.2.7.4.a check
+                    if(module.getFunction() == 0){
+                        zeroFunctionObdModules.add(module);
+
+                    }
+                });
+
+        //Get zero function module address (should only be one)
+        int function0SourceAddress = zeroFunctionObdModules.isEmpty() ? -1
+                : zeroFunctionObdModules.get(0).getSourceAddress();
+
+        //Get the packets from the global response for the zero function module
+        List<ComponentIdentificationPacket> zeroFunctionPackets = globalPackets.stream()
```

Comment:
This can find the first zero function packet ( you can merge this together with the code below)